src/hash_table/hash_table.rs

```diff
@@ -0,0 +1,3132 @@
+// Copyright 2020 Amari Robinson
+//
+// Licensed under the Apache License, Version 2.0, <LICENSE-APACHE or
+// http://apache.org/licenses/LICENSE-2.0> or the MIT license <LICENSE-MIT or
+// http://opensource.org/licenses/MIT>, at your option. This file may not be
+// copied, modified, or distributed except according to those terms.
+
+use core::borrow::Borrow;
+use core::fmt;
+use core::iter::{ExactSizeIterator, Extend, FromIterator, FusedIterator, IntoIterator};
+use core::ops::Index;
+
+use crate::pointer_ops::PointerOps;
+use crate::unchecked_option::UncheckedOptionExt;
+
+use crate::hash_table::array::Array;
+use crate::hash_table::bucket_ops::BucketOps;
+use crate::hash_table::hash_ops::HashOps;
+
+// =============================================================================
+// HashTableAdapter
+// =============================================================================
+
+/// An adapter trait which allows a type to be inserted into an intrusive hash table.
+///
+/// `PointerOps` implements the collection-specific pointer conversions which
+/// allow an object to be inserted into an intrusive hash table.
+/// `PointerOps` type may be stateful, allowing custom pointer types.
+///
+/// `BucketOps` implements the collection-specific operations which
+/// allow an object to be inserted into an intrusive hash table bucket.
+/// `BucketOps` type may be stateful, allowing custom bucket metadata without
+/// modifying the underlying container. (E.g. Bitset for tracking non-empty buckets.)
+///
+/// `KeyOps` implements the collection-specific key operations which
+/// allow an object to be inserted into an intrusive hash table.
+///
+/// `HashOps` implements the collection-specific hashing operations which
+/// allow an object to be inserted into an intrusive hash table.
+///
+/// A single object type may have multiple adapters, which allows it to be part
+/// of multiple intrusive collections simultaneously.
+///
+/// It is also possible to create stateful adapters.
+/// This allows links and containers to be separated and avoids the need for objects to be modified to
+/// contain a link.
+pub trait HashTableAdapter {
```

Comment:
`HashTableAdapter` can be simplified. Remove `PointerOps` and `BucketOps`, the adapter only needs to handle key equality and key hashing. Everything else is handled by the `BucketOps` from the `B` array parameter to `HashTable`.